File: problems/2022_09_22/0347/Solution.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        if (k == nums.length) {
            return nums;
        }

        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        Queue<Integer> heap = new PriorityQueue<>((n1, n2) -> map.get(n1) - map.get(n2));

        for (int num : map.keySet()) {
            heap.add(num);
            if (heap.size() > k) {
                heap.poll();
            }
        }

        int[] top = new int[k];
        for (int i = k - 1; i >= 0; --i) {
            top[i] = heap.poll();
        }

        return top;
    }
}

File: problems/2022_09_22/0347/Solution.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        countMap = {}

        for i in nums:
            if i in countMap.keys():
                countMap[i] += 1
            else:
                countMap[i] = 1

        countTuple = list(countMap.items())
        countTuple.sort(key=lambda x: -x[1])

        count = 0
        res = []
        while count < k and count < len(countTuple):
            res.append(countTuple[count][0])
            count += 1

        return res
